File: 面经150/12.图/133.克隆图.py
#
# @lc app=leetcode.cn id=133 lang=python3
#
# [133] 克隆图
#
from typing import Optional
from collections import *
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not  None else []
# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None
        q = deque([node])
        vis = {}
        while q:
            tmp = q.popleft()
            if tmp not in vis:
                vis[tmp] = Node(tmp.val)
            for y in tmp.neighbors:
                if y not in vis:
                    vis[y] = Node(y.val)
                    q.append(y)
                vis[tmp].neighbors.append(vis[y])
        return vis[node]
# @lc code=end

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None
        vis = {}
        def dfs(x):
            if x in vis:
                return vis[x]
            tmp = Node(x.val)
            vis[x] = tmp
            for y in x.neighbors:
                tmp.neighbors.append(dfs(y))
            return tmp
        return dfs(node)

File: 面经150/面经150p2/133.克隆图.cpp
/*
 * @lc app=leetcode.cn id=133 lang=cpp
 *
 * [133] 克隆图
 */

// @lc code=start


#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (! node) return NULL;
        queue<Node *> q;
        q.push(node);
        unordered_map<Node*, Node*> hash;
        while (q.size()) {
            Node *tmp = q.front(); q.pop();
            if (! hash.contains(tmp)) hash[tmp] = new Node(tmp->val);
            for (auto& y : tmp->neighbors) {
                if (! hash.contains(y)) {
                    hash[y] = new Node(y->val);
                    q.push(y);
                }
                hash[tmp]->neighbors.pb(hash[y]);
            }
        }
        return hash[node];
    }
};
// @lc code=end

